apps/arweave/src/ar_http_iface_server.erl

```diff
@@ -95,8 +94,181 @@ start_http_iface_listener(Config) ->
+-spec terminate_connections(Delay) -> Return when
+	Delay :: pos_integer(),
+	Return :: [{pid(), reference()}, ...].
+
+terminate_connections(Delay) ->
+	Processes = ranch:procs(ar_http_iface_listener, connections),
+	[ terminate_connection(P, Delay) || P <- Processes ].
+
+%%--------------------------------------------------------------------
+%% @doc terminate a ranch/cowboy connection. this follows the drain
+%%      procedure explaining in ranch and cowboy documentation.
+%% @end
+%%--------------------------------------------------------------------
+-spec terminate_connection(Connection, Delay) -> Return when
+	Connection :: ranch:ref(),
+	Delay :: pos_integer(),
+	Return :: pid().
+
+terminate_connection(Connection, Delay) ->
+	spawn_monitor(fun() ->
+		terminate_connection_init(Connection, Delay)
+	end).
+
+%%--------------------------------------------------------------------
+%% @hidden
+%% @doc terminate connection init procedure.
+%% @end
+%%--------------------------------------------------------------------
+-spec terminate_connection_init(Connection, Delay) -> Return when
+	Connection :: ranch:ref(),
+	Delay :: pos_integer(),
+	Return :: pid().
+
+terminate_connection_init(Connection, Delay) ->
+	% let monitor the connection first. a ranch connection is a process
+	% used as "interface" to a socket. the procedure must know if the
+	% process is still alive.
+	_Ref = erlang:monitor(process, Connection),
+
+	% it seems ranch_tcp does not have a way to "kill" or "close" the
+	% connection from the process. The port needs to be extracted, to
+	% accomplish that, one can check using process_info/2 functions or
+	% extract the process state with sys:get_state/1.
+	{links, Links} = erlang:process_info(Connection, links),
+
+	% in normal situation, a ranch connection process must have only one
+	% port (socket). If it's not the case, this is abnormal.
+	[Socket|_] = [ P || P <- Links, is_port(P) ],
+
+	% let check if information about the sockets can be retrieved, in this
+	% case, the address/port is the information required. It also check if
+	% the socket is still active. If an exception is returned, we assume
+	% the socket is already down.
+	try inet:peername(Socket) of
+		{ok, {PeerAddress, PeerPort}} ->
+			% let  prepare the state used for the termination loop procedure.
+			Peer = string:join([inet:ntoa(PeerAddress), integer_to_list(PeerPort)], ":"),
+			State = #{
+				socket => Socket,
+				peer => Peer,
+				delay => Delay,
+				connection => Connection
+			},
+
+			% now everything is ready, the socket is set in read-only mode. the
+			% remote peer will receive this information and will try to fetch
+			% the last piece of data from the buffer. It does not guarantee the
+			% socket will be closed, in particular if the connection is slow.
+			logger:warning(#{ connection => Peer, socket => Connection, reason => "shutdown", msg => "shutdown (read-only)"}),
```

Comment:
Any way to use `?LOG_WARNING` for now? to be consistent with the rest of the code